compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -753,10 +753,10 @@ class Typer extends Namer
           if (ctx.mode.is(Mode.QuotedPattern)) pt.translateFromRepeated(toArray = false, translateWildcard = true)
           else pt.translateFromRepeated(toArray = false, translateWildcard = true) |
                pt.translateFromRepeated(toArray = true,  translateWildcard = true)
-        val tpdExpr = typedExpr(tree.expr, ptArg)
         val expr1 = typedExpr(tree.expr, ptArg)
-        val fromCls = if expr1.tpe.derivesFrom(defn.ArrayClass) then defn.ArrayClass else defn.SeqClass
-        val tpt1 = TypeTree(expr1.tpe.widen.translateToRepeated(fromCls)).withSpan(tree.tpt.span)
+        val exprTpeWiden = expr1.tpe.widenUnion
```

Comment:
Any reason to call widenUnion and not stripNull?

Reply:
The `stripNull` can only handle the `| Null` case. I think this issue is for all union types?

Reply:
If both types extend Array, then their union does too and `derivesFrom` handle that, the only thing it doesn't handle is bottom types as per its documentation (and Null is considered a "bottom type" here)

Reply:
If the usage of Null becomes pervasive, maybe derivesFrom should be tweaked to handle it?

Reply:
I prefer to modify the `derivesFrom`. Is it ok to change it in this way? This will case `NullType.derivesFrom(cls)` to be `true` for all class symbol.

```scala
    /** Is this type an instance of a subclass of the given class `cls`? */
    final def derivesFrom(cls: Symbol)(using Context): Boolean = {
      def loop(tp: Type): Boolean = tp match {
        case tp: TypeRef =>
          if !ctx.explicitNulls && tp.isNullType then true
          else
            val sym = tp.symbol
            if (sym.isClass) sym.derivesFrom(cls) else loop(tp.superType)
        ......
```

Reply:
You'd need to check that `cls` itself derivesFrom AnyRef,  this change seems fine to me, but you'd have to inspect all usages of `derivesFrom` to make sure they don't rely on Null not counting /cc @odersky.

Reply:
Perhaps a less radical change would be to tweak the `OrType` branch of `derivesFrom`, allowing `X | Null` and `X | Nothing` to stand in for `X`| is less radical than also allowing `Null` and `Nothing`